File: java-samples/encumbrance-pawn-shop/workflows/src/main/java/com/r3/developers/samples/encumbrance/workflows/RequestLoanFlow.java
package com.r3.developers.samples.encumbrance.workflows;

import com.r3.developers.samples.encumbrance.contracts.AssetContract;
import com.r3.developers.samples.encumbrance.contracts.LoanContract;
import com.r3.developers.samples.encumbrance.states.Asset;
import com.r3.developers.samples.encumbrance.states.Loan;
import com.r3.developers.samples.encumbrance.states.Member;
import net.corda.v5.application.flows.ClientRequestBody;
import net.corda.v5.application.flows.ClientStartableFlow;
import net.corda.v5.application.flows.CordaInject;
import net.corda.v5.application.flows.InitiatingFlow;
import net.corda.v5.application.marshalling.JsonMarshallingService;
import net.corda.v5.application.membership.MemberLookup;
import net.corda.v5.application.messaging.FlowMessaging;
import net.corda.v5.base.annotations.Suspendable;
import net.corda.v5.base.exceptions.CordaRuntimeException;
import net.corda.v5.base.types.MemberX500Name;
import net.corda.v5.ledger.common.NotaryLookup;
import net.corda.v5.ledger.utxo.StateAndRef;
import net.corda.v5.ledger.utxo.UtxoLedgerService;
import net.corda.v5.ledger.utxo.transaction.UtxoSignedTransaction;
import net.corda.v5.ledger.utxo.transaction.UtxoTransactionBuilder;
import net.corda.v5.membership.MemberInfo;
import net.corda.v5.membership.NotaryInfo;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.time.Instant;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@InitiatingFlow(protocol = "issue-loan")
public class RequestLoanFlow implements ClientStartableFlow {

    private final static Logger log =  LoggerFactory.getLogger(RequestLoanFlow.class);
    @CordaInject
    private JsonMarshallingService jsonMarshallingService;
    @CordaInject
    private MemberLookup memberLookup;
    @CordaInject
    private NotaryLookup notaryLookup;
    @CordaInject
    private UtxoLedgerService ledgerService;
    @CordaInject
    private FlowMessaging flowMessaging;

    @NotNull
    @Override
    @Suspendable
    public String call(@NotNull ClientRequestBody requestBody) {
        try{
            RequestLoanFlowArgs flowArgs = requestBody.getRequestBodyAs(jsonMarshallingService, RequestLoanFlowArgs.class);
            MemberInfo myInfo = memberLookup.myInfo();

            MemberInfo memberInfo = memberLookup.lookup(MemberX500Name.parse(flowArgs.getLender()));
            Member lender = new Member(memberInfo.getName(), memberInfo.getLedgerKeys().get(0));

            List<StateAndRef<Asset>> filteredAssetStateAndRefs =
                    ledgerService.findUnconsumedStatesByType(Asset.class).stream().filter(
                        it -> it.getState().getContractState().getAssetId().equals(flowArgs.getCollateral())
                    ).collect(Collectors.toList());
            if (filteredAssetStateAndRefs.size() != 1) {
                throw new CordaRuntimeException(
                        "Multiple or zero Asset states with id " + flowArgs.getCollateral() + " found");
            }

            StateAndRef<Asset> assetStateAndRef = filteredAssetStateAndRefs.get(0);

            Loan loan = new Loan(
                    UUID.randomUUID().toString(),
                    new Member(lender.getName(), lender.getLedgerKey()),
                    new Member(myInfo.getName(), myInfo.getLedgerKeys().get(0)),
                    flowArgs.getLoanAmount(), flowArgs.getCollateral()
            );

            Asset collateral = new Asset(
                    assetStateAndRef.getState().getContractState().getOwner(),
                    assetStateAndRef.getState().getContractState().getAssetName(),
                    assetStateAndRef.getState().getContractState().getAssetId(),
                    Arrays.asList(
                            assetStateAndRef.getState().getContractState().getOwner().getLedgerKey(),
                            myInfo.getLedgerKeys().get(0)
                    )
            );

            MemberX500Name notaryName = assetStateAndRef.getState().getNotaryName();
            UtxoTransactionBuilder transactionBuilder = ledgerService.createTransactionBuilder()
                    .setNotary(notaryName)
                    .setTimeWindowBetween(Instant.now(), Instant.now().plusMillis(Duration.ofMinutes(5).toMillis()))
                    .addInputStates(assetStateAndRef.getRef())
                    .addEncumberedOutputStates("loan-" + loan.getLoanId(), loan, collateral)
                    .addCommand(new LoanContract.Issue())
                    .addCommand(new AssetContract.Lock())
                    .addSignatories(loan.getLender().getLedgerKey(), loan.getBorrower().getLedgerKey());
            UtxoSignedTransaction signedTransaction = transactionBuilder.toSignedTransaction();
            UtxoSignedTransaction finalizedTransaction =
                    ledgerService.finalize(signedTransaction,
                            Collections.singletonList(flowMessaging.initiateFlow(lender.getName()))).getTransaction();

            return finalizedTransaction.getId().toString();
        }catch (Exception e){
            log.warn("Failed to process flow for request body " + requestBody + " because: " + e.getMessage());
        throw new CordaRuntimeException(e.getMessage());
        }
    }
}

/* Example JSON to put into REST-API POST requestBody
{
  "clientRequestId": "request-loan",
  "flowClassName": "com.r3.developers.samples.encumbrance.workflows.RequestLoanFlow",
  "requestBody": {
    "lender": "CN=Bob, OU=Test Dept, O=R3, L=London, C=GB",
    "loanAmount": 1000,
    "collateral": "b4beb3d5-faae-4146-ab77-34e1c59f3ee9"
  }
}
*/

File: kotlin-samples/encumbrance-pawn-shop/workflows/src/main/kotlin/com/r3/developers/samples/encumbrance/workflows/RequestLoanFlow.kt
package com.r3.developers.samples.encumbrance.workflows

import com.r3.developers.samples.encumbrance.contracts.AssetContract
import com.r3.developers.samples.encumbrance.contracts.LoanContract
import com.r3.developers.samples.encumbrance.states.Asset
import com.r3.developers.samples.encumbrance.states.Loan
import com.r3.developers.samples.encumbrance.states.Member
import net.corda.v5.application.flows.*
import net.corda.v5.application.marshalling.JsonMarshallingService
import net.corda.v5.application.membership.MemberLookup
import net.corda.v5.application.messaging.FlowMessaging
import net.corda.v5.application.messaging.FlowSession
import net.corda.v5.base.annotations.Suspendable
import net.corda.v5.base.exceptions.CordaRuntimeException
import net.corda.v5.base.types.MemberX500Name
import net.corda.v5.ledger.utxo.UtxoLedgerService
import net.corda.v5.ledger.utxo.transaction.UtxoSignedTransaction
import org.slf4j.LoggerFactory
import java.time.Duration
import java.time.Instant
import java.util.*


data class RequestLoanFlowArgs(val lender: String, val loanAmount : Int, val collateral: String )

@InitiatingFlow(protocol = "issue-loan")
class RequestLoanFlow : ClientStartableFlow {
    private companion object {
        val log = LoggerFactory.getLogger(this::class.java.enclosingClass)
    }

    @CordaInject
    lateinit var jsonMarshallingService: JsonMarshallingService

    @CordaInject
    lateinit var memberLookup: MemberLookup

    // Injects the UtxoLedgerService to enable the flow to make use of the Ledger API.
    @CordaInject
    lateinit var ledgerService: UtxoLedgerService

    @CordaInject
    lateinit var flowMessaging: FlowMessaging

    @Suspendable
    override fun call(requestBody: ClientRequestBody): String {
        try{
            val (lenderArgs, loanAmount, collateral) = requestBody.getRequestBodyAs(
                jsonMarshallingService,
                RequestLoanFlowArgs::class.java
            )

            val myInfo = memberLookup.myInfo()
            val otherMember = memberLookup.lookup(MemberX500Name.parse(lenderArgs)) ?: throw CordaRuntimeException("MemberLookup can't find otherMember specified in flow arguments.")
            val lender = Member(otherMember.name,otherMember.ledgerKeys[0])

            val stateAndRef = ledgerService.findUnconsumedStatesByType(Asset::class.java).singleOrNull {
                it.state.contractState.assetId == collateral
            } ?: throw CordaRuntimeException("Multiple or zero Asset states with id ${collateral} found.")

            val input = stateAndRef.state.contractState

            val loan = Loan(
                UUID.randomUUID().toString(),
                Member(lender.name,lender.ledgerKey),
                Member(myInfo.name,myInfo.ledgerKeys[0]),
                loanAmount,
                collateral
            )

            val collateralAsset = Asset(
                input.owner,
                input.assetName,
                input.assetId,
                listOf(input.owner.ledgerKey,myInfo.ledgerKeys[0])
            )

            val notaryName = stateAndRef.state.notaryName
            val txBuilder= ledgerService.createTransactionBuilder()
                .setNotary(notaryName)
                .setTimeWindowBetween(Instant.now(), Instant.now().plusMillis(Duration.ofDays(1).toMillis()))
                .addInputState(stateAndRef.ref)
                .addEncumberedOutputStates("loan-" + loan.loanId, loan, collateralAsset)
                .addCommand(LoanContract.LoanCommands.Issue())
                .addCommand(AssetContract.AssetCommands.Lock())
                .addSignatories(loan.lender.ledgerKey, loan.borrower.ledgerKey)

            val signedTransaction: UtxoSignedTransaction = txBuilder.toSignedTransaction()
            val finalizedTransaction =
                ledgerService.finalize(signedTransaction, listOf(flowMessaging.initiateFlow(lender.name))).transaction

            return finalizedTransaction.id.toString()
        }catch (e: Exception) {
            log.warn("Failed to process utxo flow for request body '$requestBody' because:'${e.message}'")
            throw e
        }
    }
}

@InitiatedBy(protocol = "issue-loan")
class RequestLoanFlowResponder : ResponderFlow {

    private companion object {
        val log = LoggerFactory.getLogger(this::class.java.enclosingClass)
    }

    @CordaInject
    lateinit var ledgerService: UtxoLedgerService

    @Suspendable
    override fun call(session: FlowSession) {
        log.info("RequestLoanFlowResponder.call() called")

        try {
            val finalizedSignedTransaction = ledgerService.receiveFinality(session) { ledgerTransaction -> }
            log.info("Finished responder flow - ${finalizedSignedTransaction.transaction.id}")
        }
        // Soft fails the flow and log the exception.
        catch (e: Exception) {
            log.warn("Exceptionally finished responder flow", e)
        }

    }
}
/*
{
  "clientRequestId": "request-loan",
  "flowClassName": "com.r3.developers.samples.encumbrance.workflows.RequestLoanFlow",
  "requestBody": {
     "lender": "CN=Bob, OU=Test Dept, O=R3, L=London, C=GB",
     "loanAmount": 1000,
     "collateral": "<asset-id>" // Check Viewing Data in the Vault Section on get this
  }
}
 */